src/LtiMessageLaunch.php

```diff
@@ -461,25 +512,24 @@ protected function validateJwtSignature()
         return $this;
     }
 
-    protected function validateDeployment()
+    protected function validateDeployment(): self
     {
         if (!isset($this->jwt['body'][LtiConstants::DEPLOYMENT_ID])) {
             throw new LtiException(static::ERR_MISSING_DEPLOYEMENT_ID);
         }
 
         // Find deployment.
-        $client_id = is_array($this->jwt['body']['aud']) ? $this->jwt['body']['aud'][0] : $this->jwt['body']['aud'];
-        $deployment = $this->db->findDeployment($this->jwt['body']['iss'], $this->jwt['body'][LtiConstants::DEPLOYMENT_ID], $client_id);
+        $client_id = $this->getAud();
+        $this->deployment = $this->db->findDeployment($this->jwt['body']['iss'], $this->jwt['body'][LtiConstants::DEPLOYMENT_ID], $client_id);
 
-        if (empty($deployment)) {
-            // deployment not recognized.
+        if (!isset($this->deployment) && !$this->canMigrate()) {
```

Comment:
@snake @erikdonohoo before I merge this, I just want to verify a few things:

1) Is this logic `canMigrate()` accurate here? Or should it be `shouldMigrate()`?
2) I've currently got `matchingLti1p1KeyExists()` marked private. Should it be public? I wasn't sure if you would need it as part of checking `$this->db->shouldMigrate()`

Reply:
1. For our case, shouldMigrate() is really most useful when we've already passed validation of the launch data (i.e. deployment is found and valid). At this point, shouldMigrate() allows check whether we've migrated already, and also would allow us to quickly skip migration again on subsequent launches (because, of course, migrate() is still being called on every launch). So, for us, what you've got works fine (as would shouldMigrate()) since we're dealing with the case where we DO have a known deployment id. I think it's Erik's scenario that's more prevalent though and which really hits this conditional. In that case, if you don't have a deployment mapped and are unable to migrate (because of an incompatible DB type), you'd want to fail here. 

I can't think of what calling shouldMigrate() would add. It seems it might support the following:
- I don't know this deployment id
- I am equipped to migrate (i.e. I have an IMigrationDatabase)
- but I want to invalidate the launch before attempting migration

I can't think of a situation like that, personally. Maybe others can?

2. I think we would probably use the database method findLti1p1Keys() if we needed to do that, so no problem leaving it as is. As long as we have the launch data, and an opportunity to control whether migration happens, we're happy.